### PR TITLE
librbd: disable zero-copy writes by default

### PR DESCRIPTION
**Pull request by dillaman**
There is no guarantee that the write buffer provided by the user will remain
stable during the write. For example, the pagecache of a guest OS might
issue a writeback which is DMAed to QEMU and the pointer passed to librbd
for the IO. The guest OS is free to continue changing the affected page
since it will remain flagged as dirty, but the CRC computed by the
Ceph messenger layer will detect the change and it will result in an
endless session reset loop.

If we aren't checking the integrity of the data payload, we can ignore
the fact that the data might change since it will be the responsibility
of the user to eventually ensure all dirty data is written out.

Fixes: https://tracker.ceph.com/issues/42889
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
Are there other users which do correctly hold the memory static while it's being used in a write? Who did we add the zero-copy functionality for?
Static memory while the buffer is in use is one of the librados requirements, but this PR essentially means nobody gets to enjoy zero-copy. :(

**Comment by dillaman** [comment]
@gregsfortytwo None that I know of (that are major IOPS users). I added zero-copy writes for Octopus as part of several IOPS + CPU usage improvements. It's definitely non-ideal since this is a workload corner-case. I did test an alternative where `Objecter::_kick_requests` just invalidates the cached CRCs of the data payload to avoid the session reset endless loop, and that works but seemed kludgy (i.e. the connection would still be reset at least once per non-stable write).

**Comment by gregsfortytwo** [comment]
Oh I thought this had been in place for a release or two.

Anyway, makes sense to me and you've clearly thought it through. :)

File: src/librbd/ImageCtx.cc [hunk 1]
```diff
@@ -813,6 +813,16 @@ class SafeTimerSingleton : public SafeTimer {
       RBD_QOS_WRITE_BPS_THROTTLE,
       config.get_val<uint64_t>("rbd_qos_write_bps_limit"),
       config.get_val<uint64_t>("rbd_qos_write_bps_burst"));
+
+    // the write buffer might not be stable so we cannot zero-copy if the
+    // data frame is being checked for integrity
+    if (!disable_zero_copy &&
+        (config.get_val<bool>("ms_crc_data") ||
+         config.get_val<bool>("ms_msgr2_sign_messages") ||
+         config.get_val<bool>("ms_msgr2_encrypt_messages"))) {
```
Comment (idryomov) [inline src/librbd/ImageCtx.cc]: ms_msgr2_{sign,encrypt}_messages are outdated, I've submitted https://github.com/ceph/ceph/pull/31818 to remove them.  The option that is actually used is ms_client_mode, a list of modes in order of preference.

I'm not sure we should parse it in librbd though, perhaps there is a way we can query the negotiated mode, something like AuthConnectionMeta::con_mode?  @liewegas?

Comment (dillaman) [reply src/librbd/ImageCtx.cc]: > I'm not sure we should parse it in librbd though, perhaps there is a way we can query the negotiated mode, something like AuthConnectionMeta::con_mode? @liewegas?

That wouldn't work since nothing would stop one OSD connection from being encrypted and another not if mid-upgrade. Therefore, it would need to just disable zero-copy if enabled in the `ms_client_mode`. i.e. the only way to get zero-copy writes would be in benchmark tests when all these things are turned off to get snazzy numbers.

Comment (idryomov) [reply src/librbd/ImageCtx.cc]: Yeah, not sure what I was thinking.  Checking pretty much anything on a connection is racy, I should know as I've run into it more than once in the kernel...

Comment (dillaman) [reply src/librbd/ImageCtx.cc]: I could tweak the buffer::list/ptr/raw to add the concept along the lines of `make_stable` (like the old `make_shareable`) which could be invoked down in the msgr layer before writing the payload if CRCs/encryption/etc are enabled. librbd could then flag zero-copy buffer::raw as unstable and delay the copy until the last possible moment.

Comment (dillaman) [reply src/librbd/ImageCtx.cc]: ... after a quick scan of the V2 protocol code, it seems like you are either in secure more or CRC mode w/ no option to disable both, so basically we would always need to disable zero-copy writes.

Comment (dillaman) [reply src/librbd/ImageCtx.cc]: I just added a new config option to let clients opt-in if they know they will not issue unstable writes.

**Comment by idryomov** [comment]
"guarentee" in the commit message is probably a typo.  While at it, I think it's worth to expand on why qemu doesn't guarantee stable writes -- basically just copy "The guest OS is free to ..." sentence from the PR title because this bit isn't obvious.

**Comment by idryomov** [comment]
FWIW one instance of "guarentee" in still there, but I'm not sure if it's a real typo.

**Comment by dillaman** [comment]
jenkins test make check

**Comment by dillaman** [comment]
(unrelated run-tox-python-common failure)

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2019-12-09_08:59:16-rbd-wip-jd-testing-distro-basic-smithi/
